homeassistant/loader.py

```diff
@@ -1441,6 +1419,227 @@ async def async_get_integrations(
+    """Resolve all dependencies, including after_dependencies, for integrations.
+
+    Detects circular dependencies and missing integrations.
+    """
+    resolved: dict[str, set[str] | Exception] = {}
+
+    async def task(itg: Integration) -> set[str] | None:
+        if ignore_exceptions:
+            return await _do_resolve_dependencies(
+                itg,
+                cache=_ResolveDependenciesCache.from_dict(resolved),
+                possible_after_dependencies=possible_after_dependencies,
+                ignore_exceptions=ignore_exceptions,
+            )
+
+        try:
+            return await _do_resolve_dependencies(
+                itg,
+                cache=_ResolveDependenciesCache.from_dict(resolved),
+                possible_after_dependencies=possible_after_dependencies,
+                ignore_exceptions=ignore_exceptions,
+            )
+        except Exception as exc:  # noqa: BLE001
+            _log_dependencies_exceptions(itg, exc)
+            return None
+
+    resolve_dependencies_tasks = {
+        itg.domain: create_eager_task(
+            task(itg),
+            name=f"resolve after dependencies {itg.domain}",
+            loop=hass.loop,
+        )
+        for itg in integrations
+    }
+
+    result = await asyncio.gather(*resolve_dependencies_tasks.values())
+
+    return {
+        domain: deps
+        for domain, deps in zip(resolve_dependencies_tasks, result, strict=True)
+        if deps is not None
+    }
+
+
+@dataclass
+class _ResolveDependenciesCache:
+    """Cache for resolve_dependencies."""
+
+    retrieve: Callable[[Integration], set[str] | Exception | None]
+    update: Callable[[Integration, set[str] | Exception], None]
+
+    @classmethod
+    def from_dict(
+        cls,
+        cache_dict: dict[str, set[str] | Exception],
+    ) -> _ResolveDependenciesCache:
+        """Create a cache from a dictionary."""
+        return cls(
+            lambda itg: cache_dict.get(itg.domain),
+            lambda itg, all_dependencies: cache_dict.__setitem__(
+                itg.domain, all_dependencies
+            ),
+        )
+
+
+async def _do_resolve_dependencies(
+    itg: Integration,
+    *,
+    cache: _ResolveDependenciesCache | None = None,
+    possible_after_dependencies: set[str] | None | UndefinedType = UNDEFINED,
+    ignore_exceptions: bool = False,
+) -> set[str]:
+    """Recursively resolve all dependencies.
+
+    Uses `cache` to cache the results.
+
+    If `possible_after_dependencies` is not UNDEFINED, listed after dependencies are also considered.
+    If it is None, all the possible after dependencies are considered.
+
+    If `ignore_exceptions` is True, exceptions are caught and ignored and the normal resolution
+    algorithm continues. Otherwise, they are raised.
+    """
+    if cache is not None:
+        resolved = cache
+    else:
+        cache_dict: dict[str, set[str] | Exception] = {}
+        resolved = _ResolveDependenciesCache.from_dict(cache_dict)
```

Comment:
AFICT, this is only needed because tests call `_do_resolve_dependencies` directly without specifying a cache.
I'd suggest to remove it and make `cache` a non-optional parameter, and update the tests to pass in a cache.

If there's a need in the future to support cache-less calls, we can add it then.